File: src/js/assets/Parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const States_1 = require("./States");
class SQLParser {
    constructor(tokens, attrs) {
        if (tokens != null && tokens.length > 0) {
            this.tokens = tokens;
        }
        else {
            console.error("parser error: empty token");
        }
        if (attrs != null) {
            this.attr = attrs;
        }
        else {
            console.log("parser error: attributes undefined");
        }
        this.states = new Array();
        this.states[0] = States_1.startState;
        this.states[1] = States_1.state1;
        this.states[2] = States_1.state2;
        this.states[3] = States_1.state3;
        this.states[4] = States_1.state4;
        this.states[5] = States_1.state5;
        this.states[6] = States_1.state6;
        this.states[7] = States_1.state7;
        this.states[8] = States_1.endState;
        this.states[9] = States_1.errorState;
        this.state = this.states[0];
    }
    parse() {
        for (let i = 0; i < this.tokens.length; i++) {
            console.log("parsing " + this.tokens[i]);
            this.state.parse(this, this.tokens[i], this.attr);
        }
    }
    changeState(stateIndex) {
        this.state = this.states[stateIndex];
        console.log("new state" + stateIndex);
    }
    reset() {
        this.tokens = null;
        this.state = this.states[0];
    }
}
exports.SQLParser = SQLParser;

File: src/copy-fsm/command-fsm/Parser.cs
using StatePattern.States;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace StatePattern
{
    public class Parser
    {
        private State[] states;
        private State current;
        private string[][] transitions;
        private bool isAcceptable;
        private string msg;

        public Parser()
        {
            setState();
            reset();
        }

        public string parse(String toParse)
        {
            String[] parts = toParse.Split(' ');
            for(int i = 1; i < parts.Length; i++)
            {
                msg += "parsing: "+parts[i]+"\n";
                current.parse(this, parts[i]);
            }

            if(current is EndState)
            {
                isAcceptable = true;
            }
            else
            {
                isAcceptable = false;
            }

            msg += "Command accepted? " + isAcceptable;
            Task.Delay(100).ContinueWith(t => reset());
            return msg;
        }

        public void changeState(int state)
        {
           current = states[state];
           msg+= "Current state: " + current.GetType()+"\n";
        }
        
        private void setState()
        {
            Console.WriteLine("setting fsm states");
            states = new State[12];
            states[0] = new StartState();
            states[1] = new State1();
            states[2] = new State2();
            states[3] = new State3();
            states[4] = new State4();
            states[5] = new State5();
            states[6] = new State6();
            states[7] = new State7();
            states[8] = new State8();
            states[9] = new State9();
            states[10] = new EndState();
            states[11] = new ErrorState();
            transitions = new string[][]
            {
                new string[] { "0","-R","1" },
                new string[] { "0","SRC","9" },
                new string[] { "0","-H","2" },
                new string[] { "0","-L","3" },
                new string[] { "0","-P","4" },
                new string[] { "1","-H","5" },
                new string[] { "1","-L","6" },
                new string[] { "1","-P","7" },
                new string[] { "2","-R","8" },
                new string[] { "3","-R","8" },
                new string[] { "4","-R","8" },
                new string[] { "8","SRC","9" },
                new string[] { "5","SRC","9" },
                new string[] { "6","SRC","9" },
                new string[] { "7","-R","9"},
                new string[] { "9","SRC","9"},
                new string[] { "9","DST","10"},
            };
        }

        private void reset()
        {
            Console.WriteLine("reseting fsm state");
            current = states[0];
            msg = "";
        }
    }
}
